commons/src/test/java/io/aiven/kafka/connect/common/templating/TemplateTest.java

```diff
@@ -25,27 +28,155 @@
+            return true;
+        }, "Extracts " + rendered);
+    }
+}
+
 final class TemplateTest {
+
+    /**
+     * Meta-tests to ensure that the test utility detects the errors it claims and fails with meaningful messages.
+     */
+    @Test
+    void testHelper() {
+        // A basic example of a successful test
+        assertThat(Template.of("{{foo}}/{{bar}}"))
+                .satisfies(withBindings("foo", "FOO", "bar", "BAR").rendersTo("FOO/BAR"))
+                .satisfies(withInput("FOO/BAR").extracts("foo", "FOO", "bar", "BAR"));
+
+        // Failure when the test helper is misconfigured.
+        assertThatThrownBy(() -> assertThat(Template.of("{{foo}}/{{bar}}"))
+                .satisfies(withBindings("foo", "FOO", "bar").rendersTo("FOO/BAR"))).isInstanceOf(AssertionError.class)
+                .hasMessageContaining("Must set names and values in pairs")
+                .hasMessageContaining("Expecting 3 to be even");
+
+        // Failure when the instance does not rendered as expected
+        assertThatThrownBy(() -> assertThat(Template.of("{{foo}}/{{bar}}/{{baz}}"))
+                .satisfies(withBindings("foo", "FOO", "bar", "BAR").rendersTo("FOO/bar")))
+                .isInstanceOf(AssertionError.class)
+                .hasMessageContaining("expected: \"FOO/bar\"")
+                .hasMessageContaining("but was: \"FOO/BAR/{{baz}}\"");
+
+        // Failure when the extractor does not extract as expected
+        assertThatThrownBy(() -> assertThat(Template.of("{{foo}}/{{bar}}"))
+                .satisfies(withInput("FOO/BAR").extracts("foo", "foo", "bar", "bar")))
+                .isInstanceOf(AssertionError.class)
+                .hasMessageContaining("but the following map entries had different values:")
+                .hasMessageContaining("[\"foo\"=\"FOO\" (expected: \"foo\")]");
+
+        // Failure when an expected variable value is missing
+        assertThatThrownBy(
+                () -> assertThat(Template.of("{{foo}}/{{bar}}")).satisfies(withInput("FOO/BAR").extracts("foo", "FOO")))
+                .isInstanceOf(AssertionError.class)
+                .hasMessageContaining("to contain only following keys:")
+                .hasMessageContaining("[\"foo\"]");
+
+        // Failure when too many variables are expected in the returned
+        assertThatThrownBy(() -> assertThat(Template.of("{{foo}}/{{bar}}"))
+                .satisfies(withInput("FOO/BAR").extracts("foo", "FOO", "bar", "BAR", "baz", "BAZ")))
+                .isInstanceOf(AssertionError.class)
+                .hasMessageContaining("but could not find the following map entries:")
+                .hasMessageContaining("[\"baz\"=\"BAZ\"]");
+    }
+
     @Test
     void emptyString() {
-        final Template template = Template.of("");
-        assertThat(template.instance().render()).isEmpty();
+        assertThat(Template.of("")).satisfies(withBindings("foo", "FOO").rendersTo(""))
```

Comment:
There is a "special" template, where they can (but we can document against use {{timestamp:HH}}-{{timestamp:MM}}-etc

Do we facilitate that or is it just any template between {{ these and that? }}

Reply:
Yeah, this is the tricky part!  I have a couple of strategies in mind to assemble a *single* timestamp from matching *multiple* fields, but it complicated this PR -- I'm taking a look to propose a follow-up, but I'd like to do it in a separate PR.

The alternatives are:

* Match some arbitrary text inside a path like `2024/02/14/12` and either find some way to specify the `DateTimeFormatter` to interpret it correctly (and maybe the extraction pattern to be a bit more rigorous), or
* Allow matching the `yyyy`,  `mm`,  `dd`, `HH`, etc. separately and assembling them into a coherent timestamp.

WDYT?  Do you have a preference and do you think this can be done in a separate step?

At the same time, we'll also have some specific extractions that should only match digits (like partition id of course), which should be possible to specify.